monai/transforms/spatial/array.py

```diff
@@ -3230,22 +3231,22 @@ def __call__(self, array: NdarrayOrTensor):
         patched_image = np.array(patches[0])
         locations = np.array(patches[1])[:, 1:, 0]  # only keep the starting location
 
-        # Filter patches
-        if self.num_patches:
-            patched_image, locations = self.filter_count(patched_image, locations)
```

Comment:
wondering if the order of these operations matters 🤔 (filter_threshold - filter_count) vs (filter_count - filter_threshold)
I have noticed an increase in memory consumption (dataloader worker killed errors) after enabling filter_threshold on top of filter_count. 
I was able to use all available cpu cores previously... but now I have to restrict it to half the cores

Reply:
That's right, the order matters! Logically it makes sense to have threshold filtering first but it might use more memory.

Consider an image with 200 patches, where 100 are background, and a user sets `num_patches=100` and `sort_fn=None`, if we filter by count first and the apply thresholding, we only needs to apply thresholding on maximum of 100 patches but we might end up with 0 patches since all the foregrounds can be filtered by count filtering (random order).

On the other hand, when we apply thresholding first, we need to filter all the 200 patches first, and then apply count filtering (which in this example won't have any effect since there is only 100 left).

Reply:
Having said that, let me see how I can add that flexibility here.

Reply:
That would be great, thank you!

Reply:
@kenza-bouzid I added a flag `threshold_first` to handle both use cases. Could you please check with `threshold_first=False` and see if you can use all the available cpu cores? Thanks

Reply:
Oh it actually failed when I applied filter_count first then filter_threshold. 
I tried a workaround yesterday (by only changing elif by if) which forced me to limit the cpu cores, hence my question as you changed the order! Let me try with threshold first as argsort uses quicksort by default with O(n) space complexity

Reply:
I see! I enhanced the memory management by freeing up the memory for the list of extracted patch. This should help you to use more cores. Please try and let me know.

Reply:
I'm still unable to use more/all cores. 

I have encountered additional issues: 
Here's my machine specs: 8xA100 80GB RAM each, 96 cores, 1900GB CPU RAM
batch size 1, and num_patches=1700  of 224x224 (this is the max I am able to fit in my GPUs so far)
I am using cudnn.benchmark=False, cudnn.determinitic=True to ensure reproducibility

Using your current code with this setting causes: a runtime error 
`RuntimeError: Unable to find a valid cuDNN algorithm to run convolution`
I believe this usually occurs when the VRAM memory limit is hit, so I reduced the num_patches to 1500 
I still can not use all cores -> dataloader worker killed errors, reduced num_workers=6 per gpu
- For threshold_first = False: 
    Training epoch 0 takes10:07
    Training epoch 1 takes 10:01
- For threshold_first = True:
   Training epoch 0 takes 10:50
   Training epoch 1 takes 10:02

I think the difference is unsignificant, so you can remove the flag threshold_first if you prefer so

Now what is a bit puzzling for me is the 
`RuntimeError: Unable to find a valid cuDNN algorithm to run convolution`
I should be able to use the same num_patches 1700 as before.
I have also tried to set cudnn.benchmark=True, cudnn.deterministic=False (also had to reduce the num_workers to 6)
This increased the estimated total epoch time to
- 55:06 with threshold_first=False -> which now fails at step 20/103 with `CUDA out of memory`
- 61:08 with threshold_first=True -> still running so far reached step 80/103 but extremely slow 60min vs 10min (in older runs)

Reply:
Probably out of scope of this PR, but I had encountered other issues with this transform in the past.

When using num_workers>1 and pin_memory=True, training time increases exponentially over epochs 
<img width="423" alt="image" src="https://user-images.githubusercontent.com/37396332/221173718-6639f259-f2a2-4ece-9094-369980853e98.png">
<img width="429" alt="image" src="https://user-images.githubusercontent.com/37396332/221173749-4635aebc-c33a-4fa2-bc84-cd7d14c8ec9a.png">
I had profiled/timed all intermediate steps, and found out that it GridPatch was the guilty one 
<img width="466" alt="image" src="https://user-images.githubusercontent.com/37396332/221174364-580c436b-5d66-4315-812b-1beb1665f21d.png">
So I timed all intermediate step in the tranform 
 
<img width="621" alt="image" src="https://user-images.githubusercontent.com/37396332/221181608-7913298b-6bca-40df-96fd-b2137b8fe670.png">
<img width="675" alt="image" src="https://user-images.githubusercontent.com/37396332/221181800-e5ca3d07-f821-4944-a17b-121e13168e98.png">

 
<img width="689" alt="image" src="https://user-images.githubusercontent.com/37396332/221182074-4d9a3241-55e5-4be7-80e2-313dd81e36e3.png">

It turned out that the operation that was taking too long was a memory allocation by np.array
https://github.com/Project-MONAI/MONAI/blob/a2ec3752f54bfc3b40e7952234fbeb5452ed63e3/monai/transforms/spatial/array.py#L3279
I eventually fixed it by setting pin_memory=False, 
which I explain by cuda memory allocation being more expensive as it has to be allocated from the pinned memory

Note that I am dealing with particularly large slides ~50kx50k 

Any thoughts on this?

Reply:
I see! Now it makes sense. There are alternatives in MONAI that might work better for your use case but I have few questions to clarify on the situation before giving any remedy:

1. Are you loading the entire slide (50kx50k) into memory first (and then apply GridPatch)?
2. Is your loaded slide (before splitting) in numpy array or pytoch Tensor?
4. Do you have overlap between your patches?

Reply:
1. Yes, I am using LoadImaged from monai.tradforms with WSIReader, cucim backend
2. in numpy as outputted from LoadImaged 
3. I am applying some geometric augmentations (from monai) + stain normalization 
4. No 

Thanks

Reply:
Oh my bad 2. is actually a metatensor with data in uint8

Reply:
@kenza-bouzid I was not able to find the reason for the behavior that you have reported. It is very strange the time is increasing in the following epochs. Especially for convert to numpy that should take no time because there shouldn't be any copy involved.
Let's separate this issue from the counting/thresholding filtering here, so we can dig deeper and find the root cause.

Reply:
Sounds great, thanks! 
Btw I have finally been able to use all cpu cores. I restricted my batch metadata to the bare minimum which freed up some space. Thanks for your help!

Reply:
Oh and np.array does a copy, did you mean np.asarray? https://stackoverflow.com/questions/14415741/what-is-the-difference-between-np-array-and-np-asarray
I have tried changing that line https://github.com/Project-MONAI/MONAI/blob/a2ec3752f54bfc3b40e7952234fbeb5452ed63e3/monai/transforms/spatial/array.py#L3279 
by np.asarray but it didn't fix the increasing time!

Reply:
Right, I know, I meant `np.asarray` which is used in `convert_data_type(array, np.ndarray)`. I thought convert to numpy array in your profiling is referring to this.
https://github.com/Project-MONAI/MONAI/blob/a2ec3752f54bfc3b40e7952234fbeb5452ed63e3/monai/transforms/spatial/array.py#L3268

But you are referring to `np.array` in https://github.com/Project-MONAI/MONAI/blob/a2ec3752f54bfc3b40e7952234fbeb5452ed63e3/monai/transforms/spatial/array.py#L3279

Reply:
Glad that you can use all cpu cores. One additional thing is that if you don't care about having all the patches in the border (which make sense in you use case as the borders in WSI are usually representing backgrounds), you can set `pad_mode=None` which can save some space and time. I will change the default to `None` as it seems to be more useful.